### PR TITLE
ScalerTransform: micro and macro mode

### PR DESCRIPTION
**Pull request by alex-hse-repository**
**IMPORTANT: Please do not create a Pull Request without creating an issue first.**

## Before submitting (checklist)

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contribution guide](https://github.com/tinkoff-ai/etna-ts/blob/master/CONTRIBUTING.md)?
- [x] Did you check the code style? `make lint` (`poetry install -E style`).
- [ ] Did you make sure to update the docs? We use Numpy format for all the methods and classes.
- [x] Did you write any new necessary tests?
- [x] Did you check that your code passes the unit tests `pytest tests/` ? 
- [x] Did you add your new functionality to the docs?
- [ ] Did you update the [CHANGELOG](https://github.com/tinkoff-ai/etna-ts/blob/master/CHANGELOG.md)?

## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Proposed Changes

  - Add micro and macro mode for ScalerTransform

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Closing issues

Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes (if such).

### CODE REVIEW

File: etna/transforms/scalers.py [hunk 1]
```diff
@@ -37,11 +39,16 @@ def __init__(
             if True, center the data before scaling.
         with_std:
             if True, scale the data to unit standard deviation.
+        mode:
+            "macro" or "per-segment", way to transform features over segments.
+            If "macro", transforms features globally, gluing the corresponding ones for all segments.
+            If "per-segment", transforms features for each segment separately.
```
Comment (Mr-Geekman) [inline etna/transforms/scalers.py:45]: May be we should add Raises block here.

Comment (alex-hse-repository) [reply etna/transforms/scalers.py:45]: We don't have Raises block in the similar cases in `_OneSegmentTimeSeriesImputerTransform` or `Metric`

File: etna/transforms/scalers.py [hunk 2]
```diff
@@ -23,6 +24,7 @@ def __init__(
         inplace: bool = True,
         with_mean: bool = True,
         with_std: bool = True,
+        mode: str = TransformMode.per_segment,
```
Comment (martins0n) [inline etna/transforms/scalers.py:27]: type TransformMode

File: etna/transforms/sklearn.py [hunk 3]
```diff
@@ -1,20 +1,37 @@
...
 from etna.transforms.base import Transform
 
 
+class TransformMode(str, Enum):
+    """Enum for different metric aggregation modes."""
+
+    macro = "macro"
+    per_segment = "per-segment"
+
+    @classmethod
```
Comment (martins0n) [inline etna/transforms/sklearn.py]: We should remove this.
Maybe we will add this with mixin later for every enum.

File: etna/transforms/sklearn.py [hunk 4]
```diff
@@ -99,10 +131,31 @@ def inverse_transform(self, df: pd.DataFrame) -> pd.DataFrame:
...
+                transformed = self._inverse_reshape(df, transformed)
+
             df.loc[:, pd.IndexSlice[:, self.in_column]] = transformed
         return df
 
+    def _reshape(self, df: pd.DataFrame) -> pd.DataFrame:
+        segments = sorted(set(df.columns.get_level_values("segment")))
+        x = df.loc[:, pd.IndexSlice[:, self.in_column]]
+        x = pd.concat([x[segment] for segment in segments]).values
+        return x
```
Comment (martins0n) [inline etna/transforms/sklearn.py:149]: `x` - is numpy array, isnt it?

File: etna/transforms/scalers.py (hunk 1 above)
Comment (alex-hse-repository) [reply etna/transforms/scalers.py:45]: Done

File: etna/transforms/scalers.py [hunk 5]
```diff
@@ -37,11 +39,21 @@ def __init__(
...
             if True, scale the data to unit standard deviation.
+        mode:
+            "macro" or "per-segment", way to transform features over segments.
+            If "macro", transforms features globally, gluing the corresponding ones for all segments.
+            If "per-segment", transforms features for each segment separately.
+
+        Raises
+        ------
+        ValueError:
+            if incorrect strategy given
```
Comment (Mr-Geekman) [inline etna/transforms/scalers.py:50]: strategy -> mode

File: etna/transforms/scalers.py [hunk 6]
```diff
@@ -82,6 +95,15 @@ def __init__(
...
             the dataset will be scaled up.
+        mode:
+            "macro" or "per-segment", way to transform features over segments.
+            If "macro", transforms features globally, gluing the corresponding ones for all segments.
+            If "per-segment", transforms features for each segment separately.
+
+        Raises
+        ------
+        ValueError:
+            if incorrect strategy given
```
Comment (Mr-Geekman) [inline etna/transforms/scalers.py:106]: strategy -> mode

Comment (alex-hse-repository) [reply etna/transforms/scalers.py]: Done

**Review by Mr-Geekman — APPROVED** [review]

File: etna/transforms/sklearn.py [hunk 7]
```diff
@@ -99,10 +131,31 @@ def inverse_transform(self, df: pd.DataFrame) -> pd.DataFrame:
...
-            transformed = self.transformer.inverse_transform(X=x)
+            if self.mode == TransformMode.per_segment:
+                x = df.loc[:, pd.IndexSlice[:, self.in_column]].values
+                transformed = self.transformer.inverse_transform(X=x)
+
+            elif self.mode == TransformMode.macro:
+                x = self._reshape(df)
+                transformed = self.transformer.inverse_transform(X=x)
+                transformed = self._inverse_reshape(df, transformed)
+
```
Comment (martins0n) [inline etna/transforms/sklearn.py:141]: It's not great finalize without else statement -- this is kind of pattern matching and we should raise Exception or doing something for other possible cases. 
The probablity of self.mode not equal to something from enum is mostly equal to zero. But it can be, for example you change mode by hand in tests and you could pass everything. 
So we should raise ValueError in `else` block

**Review by martins0n — CHANGES_REQUESTED** [review]
We should add the same for power transforms.

File: etna/transforms/scalers.py (hunk 2 above)
Comment (alex-hse-repository) [reply etna/transforms/scalers.py:27]: Here we pass the sting, not the TransformMode

File: etna/transforms/sklearn.py (hunk 3 above)
Comment (alex-hse-repository) [reply etna/transforms/sklearn.py]: Done

File: etna/transforms/sklearn.py (hunk 4 above)
Comment (alex-hse-repository) [reply etna/transforms/sklearn.py:149]: Done

File: etna/transforms/sklearn.py (hunk 7 above)
Comment (alex-hse-repository) [reply etna/transforms/sklearn.py:141]: Done

File: etna/transforms/scalers.py (hunk 2 above)
Comment (alex-hse-repository) [reply etna/transforms/scalers.py:27]: Done

Comment (martins0n) [reply etna/transforms/scalers.py:27]:
```
>>> from etna.transforms.sklearn import TransformMode
>>> TransformMode.per_segment
<TransformMode.per_segment: 'per-segment'>
>>> type(TransformMode.per_segment)
<enum 'TransformMode'>
```
?

Comment (alex-hse-repository) [reply etna/transforms/scalers.py]: There should be "per-segment" instead of TransformMode.per_segment)

**Review by martins0n — APPROVED** [review]
👍
